docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,72 @@
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* It can be tied to Thanos.
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result.
+
+Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.
+
+## Alternatives
+
+* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).
+
+## Work Plan
+
+* Implement a new flag `--rule` in Thanos Query which registers RulesAPI endpoints.
+* Implement RulesAPI backends in sidecar, query, rule.
+* Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.
+* Feature branch: https://github.com/thanos-io/thanos/pull/2200
```

Comment:
It's not a plan item if you had already started it, right? So I think this should be removed as well.

Reply:
Still it was planned, so let's link it somehow, no? (:

Reply:
agreed, we should keep the link to the WIP PR somewhere :thinking:

Reply:
Okay :+1: at least it's not at the beginning :smile: